// braces/lib/constants.js
'use strict';

module.exports = {
  MAX_LENGTH: 1024 * 64,
  SPECIAL_CHARS: {
    DIGITS: {
      ZERO: '0',
      NINE: '9'
    },
    ALPHABET: {
      UPPERCASE_A: 'A',
      LOWERCASE_A: 'a',
      UPPERCASE_Z: 'Z',
      LOWERCASE_Z: 'z'
    },
    PARENTHESES: {
      LEFT: '(',
      RIGHT: ')'
    },
    ASTERISK: '*',
    NON_ALPHABETIC: {
      AMPERSAND: '&',
      AT: '@',
      BACKSLASH: '\\',
      BACKTICK: '`',
      CARRIAGE_RETURN: '\r',
      CIRCUMFLEX_ACCENT: '^',
      COLON: ':',
      COMMA: ',',
      DOLLAR: '$',
      DOT: '.',
      DOUBLE_QUOTE: '"',
      EQUAL: '=',
      EXCLAMATION_MARK: '!',
      FORM_FEED: '\f',
      FORWARD_SLASH: '/',
      HASH: '#',
      HYPHEN_MINUS: '-',
      LEFT_ANGLE_BRACKET: '<',
      LEFT_CURLY_BRACE: '{',
      LEFT_SQUARE_BRACKET: '[',
      LINE_FEED: '\n',
      NO_BREAK_SPACE: '\u00A0',
      PERCENT: '%',
      PLUS: '+',
      QUESTION_MARK: '?',
      RIGHT_ANGLE_BRACKET: '>',
      RIGHT_CURLY_BRACE: '}',
      RIGHT_SQUARE_BRACKET: ']',
      SEMICOLON: ';',
      SINGLE_QUOTE: '\'',
      SPACE: ' ',
      TAB: '\t',
      UNDERSCORE: '_',
      VERTICAL_LINE: '|',
      ZERO_WIDTH_NOBREAK_SPACE: '\uFEFF'
    }
  }
};
